### PR TITLE
Enable user to share wallet address 

### LINKED ISSUE #17865: Share Screen -> Implement Share Wallet Address
From the designs:
https://www.figma.com/file/eDfxTa9IoaCMUy5cLTp0ys/Shell-for-Mobile?type=design&node-id=2937-643562&mode=design&t=r1UAixLCMYOEphsl-4

- Please align with @ulisesmac for the correct components to use for this as some adjustments were recently made for the QR sharing components.

Check with design of what account is to be shared here - is it the default wallet account or like the last viewed account? 

Acceptance Criteria:
- implement the UI and use real data where possible.
- Component spec test covering the screen. (entry added in status_im2/core_spec.cljs)

### PR DESCRIPTION
**Pull request by Pau1fitz**
Fixes https://github.com/status-im/status-mobile/issues/17865

Share Screen -> Implement Share Wallet Address

https://github.com/status-im/status-mobile/assets/6873736/a83ce46d-acbb-466e-af28-38a76d8f8ebe

Figma [design](https://www.figma.com/file/eDfxTa9IoaCMUy5cLTp0ys/Shell-for-Mobile?type=design&node-id=2937-665825&mode=design&t=pAmmCaduZNBaHM7B-0)

iOS
Android

Testing notes:
Navigate to the share screen
click on Wallet
user can see each wallet that to share address

### CODE REVIEW

File: src/status_im/contexts/shell/share/profile/view.cljs [hunk 1]
```diff
@@ -0,0 +1,75 @@
+(ns status-im.contexts.shell.share.profile.view
```
Comment (Pau1fitz) [inline src/status_im/contexts/shell/share/profile/view.cljs:1]: Created a new file `share/profile/view` and moved the `profile` code below, that was previously in `share/view`, to improve code organization and readability.

File: src/status_im/contexts/shell/share/view.cljs [hunk 2]
```diff
@@ -43,72 +39,6 @@
      :style  style/header-heading}
     (i18n/label :t/share)]])
 
-(defn profile-tab
```
Comment (Pau1fitz) [inline src/status_im/contexts/shell/share/view.cljs:46]: this code was moved to new file above - share/profile/view

File: src/status_im/contexts/shell/share/wallet/view.cljs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            :customization-color (:color account)
+            :emoji               (:emoji account)
+            :on-multichain-press #(reset! wallet-type :wallet-multichain)
+            :on-legacy-press     #(reset! wallet-type :wallet-legacy)
+            :on-settings-press   #(open-preferences selected-networks)}]]]
+      ))))
+
+(defn wallet-tab
+  []
+  (let [accounts (rf/sub [:wallet/accounts])
```
Comment (Pau1fitz) [inline src/status_im/contexts/shell/share/wallet/view.cljs:79]: `(rf/sub [:wallet/accounts])` also includes watched addresses. Are watched addresses intended to be displayed?

File: src/status_im/contexts/shell/share/view.cljs [hunk 4]
```diff
@@ -1,18 +1,14 @@
 (ns status-im.contexts.shell.share.view
   (:require
-    [clojure.string :as string]
```
Comment (Pau1fitz) [inline src/status_im/contexts/shell/share/view.cljs:3]: The removed code in this file was moved to the new file above - `share/profile/view`

File: src/status_im/contexts/shell/share/wallet/view.cljs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+       :snap-to-alignment         "start"
+       :snap-to-interval          (- width 30)
+       :disable-interval-momentum true
+       :scroll-event-throttle     64
+       }
+      (map-indexed
+       (fn [index account]
+         ^{:key (:address account)}
+         [wallet-qr-code-item account width index])
+       accounts)]]))
```
Comment (OmarBasem) [inline src/status_im/contexts/shell/share/wallet/view.cljs]: Better use a FlatList here.

File: src/status_im/contexts/shell/share/wallet/view.cljs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            :on-share-press      #(share-action qr-url share-title)
+            :profile-picture     nil
+            :unblur-on-android?  true
+            :full-name           (:name account)
+            :customization-color (:color account)
+            :emoji               (:emoji account)
+            :on-multichain-press #(reset! wallet-type :wallet-multichain)
+            :on-legacy-press     #(reset! wallet-type :wallet-legacy)
+            :on-settings-press   #(open-preferences selected-networks)}]]]
+      ))))
```
Comment (OmarBasem) [inline src/status_im/contexts/shell/share/wallet/view.cljs]: Trailing `)))`

File: src/status_im/contexts/shell/share/wallet/view.cljs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        width    (rf/sub [:dimensions/window-width])]
+    [rn/view
+     [rn/scroll-view
+      {:horizontal                true
+       :deceleration-rate         0.9
+       :snap-to-alignment         "start"
+       :snap-to-interval          (- width 30)
+       :disable-interval-momentum true
+       :scroll-event-throttle     64
+       }
```
Comment (OmarBasem) [inline src/status_im/contexts/shell/share/wallet/view.cljs]: Trailing `}`

File: src/status_im/contexts/shell/share/wallet/view.cljs (hunk 5 above)
Comment (siddarthkay) [reply src/status_im/contexts/shell/share/wallet/view.cljs]: Hey @OmarBasem : I think its okay to use a scroll view for a list that is not too long.
What do you think will be the maximum number of wallet accounts a user would have?

Comment (John-44) [reply src/status_im/contexts/shell/share/wallet/view.cljs]: > 

@siddarthkay there isn't currently an upper limit, but realistically around 20 would be the max anybody would have in the worst case.

File: src/status_im/contexts/shell/share/wallet/view.cljs (hunk 3 above)
Comment (xAlisher) [reply src/status_im/contexts/shell/share/wallet/view.cljs:79]: Watched addresses offer a dual purpose: they allow watching any address, as well as monitoring your own personal accounts without divulging keys. Consequently, shell will display these accounts. Pedro and Mario are planning to introduce a new component state for the "Share QR Code", which will include support for watched accounts in the near future. @pedro-et @mariocnovoa

Comment (Pau1fitz) [reply src/status_im/contexts/shell/share/wallet/view.cljs:79]: thanks @xAlisher

File: src/status_im/contexts/shell/share/wallet/view.cljs (hunk 5 above)
Comment (Pau1fitz) [reply src/status_im/contexts/shell/share/wallet/view.cljs]: I would have thought given the estimated size of the data `ScrollView` should be suffice. In general I have always used `FlatList` for larger datasets. 

Another practical point is that it is a little bit more difficult to manage state with a `FlatList` as component state is not maintained due to the components unmounting once they are off the screen and being recrated from scratch once displayed again. In this case state is required to display the `Legacy` or `Multichain` address.

If people feel strongly against this I am more than happy to implement using a `FlatList` component.

File: src/status_im/contexts/shell/share/wallet/view.cljs (hunk 7 above)
Comment (Pau1fitz) [reply src/status_im/contexts/shell/share/wallet/view.cljs]: thanks, updated 👍

File: src/status_im/contexts/shell/share/wallet/view.cljs (hunk 6 above)
Comment (Pau1fitz) [reply src/status_im/contexts/shell/share/wallet/view.cljs]: 👍

File: src/status_im/contexts/shell/share/wallet/view.cljs (hunk 5 above)
Comment (Pau1fitz) [reply src/status_im/contexts/shell/share/wallet/view.cljs]: When testing 20 items with `ScrollView` I noticed some performance issues. Will update to use `FlatList`

Comment (Pau1fitz) [reply src/status_im/contexts/shell/share/wallet/view.cljs]: I will do some performance testing with 20 items and compare both approaches side by side.

Comment (Pau1fitz) [reply src/status_im/contexts/shell/share/wallet/view.cljs]: 20 items

`FlatList`

https://github.com/status-im/status-mobile/assets/6873736/bbb2ecf7-d5ac-488c-abef-dd861e9c380b


`ScrollView`

https://github.com/status-im/status-mobile/assets/6873736/b27eed41-55de-4c54-9c97-35a04fc54ba1

@OmarBasem @siddarthkay  Looks like there is a noticeable delay using  `ScrollView`, so I have updated to use `FlatList`.

**Review by siddarthkay — APPROVED** [review]

File: src/status_im/contexts/shell/share/wallet/view.cljs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                                                  :content address}
+                               :item             {:default {:type "text"
+                                                            :content
+                                                            address}}
+                               :link-metadata    {:title share-title}}]}
+     {:title   share-title
+      :subject share-title
+      :message address})))
+
+(defn- open-preferences
```
Comment (J-Son89) [inline src/status_im/contexts/shell/share/wallet/view.cljs:32]: imo: it's better to pass the deref'd value unless you need to change it, that way the function just takes a param and doesn't need to know whether it's a ratom or not.

e.g
```
(defn- open-preferences
  [selected-networks]
  (rf/dispatch [:show-bottom-sheet
                {:theme :dark
                 :shell? true
                 :content
                 (fn []
                   [network-preferences/view
                    {:blur?             true
                     :selected-networks (set selected-networks)
                     :on-save           (fn [chain-ids]
                                          (rf/dispatch [:hide-bottom-sheet])
                                          (reset! selected-networks (map #(get utils/id->network %)
                                          
                                        
```
and then used as `#(open-preferences @selected-networks)`

Comment (Pau1fitz) [reply src/status_im/contexts/shell/share/wallet/view.cljs:32]: good shout @J-Son89, makes sense - updated 👍

**Review by OmarBasem — APPROVED** [review]
Thanks for your work @Pau1fitz !

**Comment by pavloburykh** [comment]
@Pau1fitz thanx for the PR!

Please, take a look at the issues

### ISSUE 1 Share wallet address button does not respond on IOS

Reproducible on IOS only
Tested on IOS 16.7.2, iPhone X

**Steps:** 

1. Open share wallet tab
2. Tap in share button
3. See if share button responses

**Actual result:** share button does not respond

photo_2024-01-10 16 33 25

**Expected result:** bottom sheet with receivers list should open

**Comment by pavloburykh** [comment]
### ISSUE 2 Share wallet elements are not fixed (can be tapped and moved to different directions) IOS 

Reproducible on IOS only
Tested on IOS 16.7.2, iPhone X

**Steps:** 

1. Open Share wallet tab
2. Tap QR element and try to move it. 

**Actual result:** elements are not fixed, can be moved to different sides. 


https://github.com/status-im/status-mobile/assets/97245802/3d689a8e-1ce8-4c01-a871-817b656a8023

**Expected result:** user should only be able to slide the element in case of multiple addresses.

**Comment by pavloburykh** [comment]
### ISSUE 3 Selected app is opened within Status app tap during sharing of wallet address (Android)

Tested on Android 12, Samsung Galaxy A 52

**Steps:** 

1. Open share wallet tab
2. Tap share button
3. Select app for sharing 
4. See if app is opened in the new tab

**Actual result:** selected app is opened within Status app. User needs to tap system back button to come back to Status app screens. 

photo_2024-01-10 17 07 25

**Expected result:** selected app should be opened in separate tab.

**Comment by pavloburykh** [comment]
### ISSUE 4 Background of selected wallet does not inherit selected wallet colour 

**Actual result:** background colour is always the same

photo_2024-01-10 17 12 22


**Expected result:** background color inherits selected wallet colour

[Figma](https://www.figma.com/file/eDfxTa9IoaCMUy5cLTp0ys/Shell-for-Mobile?type=design&node-id=2937-665825&mode=design&t=bZthfgXXubuoY6Xr-0)

Shell for Mobile – Figma 2024-01-10 17-09-09

**Comment by pavloburykh** [comment]
### ISSUE 5 Share wallet screen misses slider indicator

@Pau1fitz these are the new updates form design team. Would be great if we can implement it in scope of this PR. 

**Actual result:** no slider indicator

photo_2024-01-10 17 21 44

**Expected result:** slider indicator is displayed

[Figma](https://www.figma.com/file/eDfxTa9IoaCMUy5cLTp0ys/Shell-for-Mobile?type=design&node-id=2937-665825&mode=design&t=bZthfgXXubuoY6Xr-0)

Shell for Mobile – Figma 2024-01-10 17-17-43

**Comment by OmarBasem** [comment]
@pavloburykh Please ignore UI issues. We have separate issues for issues 4 and 5
https://github.com/status-im/status-mobile/issues/18452
https://github.com/status-im/status-mobile/issues/18454

**Comment by OmarBasem** [comment]
@pavloburykh regarding issue 3, 
I think this is handled by the target app. Some apps open on top of the current app (ex.: Gmail), and others open in another window (ex.: Outlook)

Update: this behavior can be overridden.

**Comment by pavloburykh** [comment]
> @pavloburykh regarding issue 3,
> This is handled by the target app. Some apps open on top of the current app (ex.: Gmail), and others open in another window (ex.: Outlook)

@OmarBasem I was comparing to sharing the profile. In case with Profile we open the app in another window. So I guess we need/can make the same implementation for wallet sharing. 


https://github.com/status-im/status-mobile/assets/97245802/5d6e00da-3c5e-4083-9eb0-09311b96b9c3

**Comment by OmarBasem** [comment]
Thanks @pavloburykh you are right. 

Do we always want to open in a new window tho? I am thinking if the current profile sharing behaviour should be on top of the current window? 🤔

**Comment by pavloburykh** [comment]
> Thanks @pavloburykh you are right.
> 
> Do we always want to open in a new window tho? I am thinking if the current profile sharing behaviour should be on top of the current window? 🤔

If we do it on top of the current window (like it is currently implemented in this PR) user stays in the selected app and is not even redirected back to Status share screen after sending wallet address. That does not look user friendly IMO. Because in this case user needs to tap system back button couple of times until he is returned to initial Share wallet screen. Also It looks weird that after sharing wallet address User can continue interacting with selected app that is opened on top of Status app window.

**Comment by J-Son89** [comment]
> > Thanks @pavloburykh you are right.
> 
> > 
> 
> > Do we always want to open in a new window tho? I am thinking if the current profile sharing behaviour should be on top of the current window? 🤔
> 
> 
> 
> If we do it on top of the current window (like it is currently implemented in this PR) user stays in the selected app and is not even redirected back to Status share screen after sending wallet address. That does not look user friendly IMO. Because in this case user needs to tap system back button couple of times until he is returned to initial Share wallet screen. Also It looks weird that after sharing wallet address User can continue interacting with selected app that is opened on top of Status app window. 

@OmarBasem, is this a common issue to the wallet share screens? If so perhaps it is beyond the scope of this pr? 🤔

**Comment by pavloburykh** [comment]
> @OmarBasem, is this a common issue to the wallet share screens? If so perhaps it is beyond the scope of this pr? 🤔

@J-Son89 if by "wallet share screen" you mean this screen below - then the answer is yes. ISSUE 1 and ISSUE 3 are common. So probably those issues can be handled separately, not in scope of this PR. 

photo_2024-01-11 15 38 22

**Comment by J-Son89** [comment]
> > @OmarBasem, is this a common issue to the wallet share screens? If so perhaps it is beyond the scope of this pr? 🤔
> 
> 
> 
> @J-Son89 if by "wallet share screen" you mean this screen below - then the answer is yes. ISSUE 1 and ISSUE 3 are common. So probably those issues can be handled separately, not in scope of this PR. 
> 
> 
> 
> photo_2024-01-11 15 38 22
> 
> 

Thanks @pavloburykh - yes exactly, I mean those screens. Ok perfect, sounds good 👌

**Comment by pavloburykh** [comment]
Okay, so do I understand correctly that among all above issues we want to fix only [ISSUE 2](https://github.com/status-im/status-mobile/pull/18424#issuecomment-1885059483) in scope of this PR? Or ISSUE 2 is also considered not to be in scope? @OmarBasem @J-Son89

**Comment by J-Son89** [comment]
> Okay, so do I understand correctly that among all above issues we want to fix only [ISSUE 2](https://github.com/status-im/status-mobile/pull/18424#issuecomment-1885059483) in scope of this PR? Or ISSUE 2 is also considered not to be in scope? @OmarBasem @J-Son89 

Oh issue 2 looks related to this pr 👌

**Comment by Pau1fitz** [comment]
@J-Son89 Issue 1 has been fixed and tested by @pavloburykh

**Comment by Pau1fitz** [comment]
@J-Son89 Issue 3 has also been fixed and tested by @pavloburykh

**Comment by pavloburykh** [comment]
Thank you @Pau1fitz 

So, here the summury:

ISSUE 1 and 3 are fixed. 
ISSUE 4 and 5 will be fixed in https://github.com/status-im/status-mobile/issues/18452 and https://github.com/status-im/status-mobile/issues/18454

ISSUE 2 is fixed partially: still can scroll QR up and down. We agreed to log this issue separately as it is device/IOS specific. 


https://github.com/status-im/status-mobile/assets/97245802/97fb8cdd-5378-4fb8-97b9-be86f408d464

PR is ready for merge.

**Comment by J-Son89** [comment]
@Pau1fitz - I think you have to sign commits for the pr to be merged https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

**Comment by Pau1fitz** [comment]
@J-Son89 I spoke to @flexsurfer about this and I am happy for a core dev to merge this on my behalf.

**Comment by J-Son89** [comment]
merged here: https://github.com/status-im/status-mobile/pull/18511

**Pull request closed without merging**